File: 品味Java的21種設計模式/dp_cpp/decorator/example2/C4.cpp
#include "C4.h"

namespace cn
{
	namespace javass
	{
		namespace dp
		{
			namespace decorator
			{
				namespace example2
				{

					void C4::setB(B *b)
					{
						this->b = b;
					}

					void C4::a1()
					{
						//转调A对象的功能
						a->a1();
					}

					void C4::b1()
					{
						//转调B对象的功能
						b->b1();
					}

					void C4::c11()
					{
						puts("now in C4.c11");
					}
				}
			}
		}
	}
}

File: 品味Java的21種設計模式/dp_java/decorator/example2/C4.java
package cn.javass.dp.decorator.example2;

public class C4 {
    //示例直接在属性上创建需要组合的对象
    private A a = new A();
    //示例通过外部传入需要组合的对象
    private B b = null;
    public void setB(B b) {
        this.b = b;
    }
    public void a1() {
        //转调A对象的功能
        a.a1();
    }
    public void b1() {
        //转调B对象的功能
        b.b1();
    }
    public void c11() {
        System.out.println("now in C4.c11");
    }
}
